scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/context/TaskContextVariableExtractor.java

```diff
@@ -36,147 +38,342 @@
-        Map<String, String> thirdPartyCredentials = forkedTaskVariablesManager.extractThirdPartyCredentials(taskContext);
+        Map<String, Serializable> variables = new HashMap<>();
+
+        try {
+            variables = extractAllVariables(taskContext, null, "");
+        } catch (IOException | ClassNotFoundException e) {
+            logger.error("Error reading variables from task context!");
+        }
+
+        Map<String, String> thirdPartyCredentials = new HashMap<>();
+
+        try {
+            thirdPartyCredentials = forkedTaskVariablesManager.extractThirdPartyCredentials(taskContext);
+        } catch (Exception e) {
+            logger.error("Error decrypting third party credentials!");
+        }
+
         HashMap<String, Serializable> systemEnvironmentVariables = new HashMap<String, Serializable>(System.getenv());
         systemEnvironmentVariables.putAll(variables);
         systemEnvironmentVariables.putAll(thirdPartyCredentials);
 
-        return VariableSubstitutor.filterAndUpdate(forkEnvironment.getSystemEnvironment(), systemEnvironmentVariables);
+        return filterAndUpdate(forkEnvironment.getSystemEnvironment(), systemEnvironmentVariables);
     }
 
-    public Map<String, Serializable> extractVariables(TaskContext taskContext, boolean useTaskVariables)
-            throws IOException, ClassNotFoundException {
-        return extractVariables(taskContext, (TaskResult) null, useTaskVariables);
-    }
-
-    private Map<String, Serializable> extractVariables(TaskContext container, TaskResult taskResult, String nodesFile,
-            boolean useTaskVariables) throws IOException, ClassNotFoundException {
-        Map<String, Serializable> variables = extractVariables(container, taskResult, useTaskVariables);
+    /**
+     * Return all variables in scope of a given taskContext.
+     *
+     * @param taskContext task context container.
+     * @return Map containing all variables resolved.
+     */
+    public Map<String, Serializable> getScopeVariables(TaskContext taskContext) {
+        Map<String, Serializable> variables = new HashMap<>();
+        Map<String, Serializable> inherited = new HashMap<>();
+        Map<String, Serializable> dictionary = new HashMap<>();
 
-        variables.put(SchedulerVars.PA_NODESNUMBER.toString(), container.getOtherNodesURLs().size() + 1);
-        variables.put(SchedulerVars.PA_NODESFILE.toString(), nodesFile);
+        try {
+            inherited.putAll(extractSystemVariables(taskContext, ""));
+            inherited.putAll(extractJobVariables(taskContext));
+            inherited.putAll(extractInheritedVariables(taskContext));
 
-        variables.put(SchedulerVars.PA_TASK_PROGRESS_FILE.toString(), container.getProgressFilePath());
+            for (TaskVariable taskVariable : taskContext.getInitializer().getTaskVariables().values()) {
+                if (!taskVariable.isJobInherited()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this conditional structure or edit its code blocks so that they're not all the same. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3923)